packages/memory-storage/src/index.ts

```diff
@@ -0,0 +1,112 @@
+    private readonly log = defaultLog.child({ prefix: 'MemoryStorage' });
+
+    constructor(options: MemoryStorageOptions = {}) {
+        s.object({
+            localDataDirectory: s.string.optional,
+        }).parse(options);
+
+        this.localDataDirectory = options.localDataDirectory ?? process.env.APIFY_LOCAL_STORAGE_DIR ?? './memory-storage';
+        this.datasetsDirectory = resolve(this.localDataDirectory, 'datasets');
+        this.keyValueStoresDirectory = resolve(this.localDataDirectory, 'key_value_stores');
+        this.requestQueuesDirectory = resolve(this.localDataDirectory, 'request_queues');
+
+        this.createWorker();
+    }
+
+    datasets(): storage.DatasetCollectionClient {
+        return new DatasetCollectionClient(this.datasetsDirectory, this);
+    }
+
+    dataset<Data extends Dictionary = Dictionary>(id: string): storage.DatasetClient<Data> {
+        s.string.parse(id);
+
+        return new DatasetClient({ id, baseStorageDirectory: this.datasetsDirectory, client: this });
+    }
+
+    keyValueStores(): storage.KeyValueStoreCollectionClient {
+        return new KeyValueStoreCollectionClient(this.keyValueStoresDirectory, this);
+    }
+
+    keyValueStore(id: string): storage.KeyValueStoreClient {
+        s.string.parse(id);
+
+        return new KeyValueStoreClient({ id, baseStorageDirectory: this.keyValueStoresDirectory, client: this });
+    }
+
+    requestQueues(): storage.RequestQueueCollectionClient {
+        return new RequestQueueCollectionClient(this.requestQueuesDirectory, this);
+    }
+
+    requestQueue(id: string, options: storage.RequestQueueOptions = {}): storage.RequestQueueClient {
+        s.string.parse(id);
+        s.object({
+            clientKey: s.string.optional,
+            timeoutSecs: s.number.optional,
+        }).parse(options);
+
+        return new RequestQueueClient({ id, baseStorageDirectory: this.requestQueuesDirectory, client: this, ...options });
+    }
+
+    private createWorker() {
+        const directoryMap = {
+            datasetsDirectory: this.datasetsDirectory,
+            keyValueStoresDirectory: this.keyValueStoresDirectory,
+            requestQueuesDirectory: this.requestQueuesDirectory,
+        };
+
+        const workerPath = resolve(__dirname, './workers/file-storage-worker.js');
+        const exists = existsSync(workerPath);
+
+        if (exists) {
+            this.fileStorageWorker = new Worker(workerPath, {
+                workerData: directoryMap,
+            });
+
+            this.fileStorageWorker.once('exit', (code) => {
+                this.log.debug(`File storage worker exited with code ${code}`);
+                this.createWorker();
+            });
+        } else {
+            this.fileStorageWorker = new FileStorageWorkerEmulator(directoryMap);
+        }
+    }
+
+    protected sendMessageToWorker(message: WorkerReceivedMessage) {
+        this.fileStorageWorker.postMessage(message);
+    }
```

Comment:
The method shouldn't be used by outside users, and typing it as public (even if annotated with `@internal`) will make it show up in intellisense. The bracket syntax is a trick that's by default in TS (and eslint is complaining because we're using the eslint dot-notation rule not typescript-eslints rule). The only reason I made this a method is to strictly type the message property. We *could* also make the worker be a singleton instead of created per storage instance, which would solve this. What do you think is the better option?